### PR TITLE
Illuminate 5.5 for next-back

### PR DESCRIPTION
**Pull request by luceos**
So here's the work progress for the upgrade to the (Laravel) Illuminate components 5.5 Long Term Support.

---

Trace of work:

- Collection `lists` modified to `pluck`
- Validation exception testing fixed
- Increased minimum constraint of PHP to 7.0, matching Illuminate framework.
- Removed PHP 5.6 and added PHP 7.2 in travis config.
- Implemented Illuminate Translator contract.

### CODE REVIEW

**Comment by tobscure** [comment]
Didn't realise 5.5 requires PHP 7... @franzliedke are we OK with this?

**Comment by Eldenroot** [comment]
@tobscure - please check this: http://php.net/supported-versions.php

Flarum should support PHP 7.0 and above. Almost 90% of webhosting offer php 7.0.

**Comment by franzliedke** [comment]
I'm fine with PHP 7.

File: src/Foundation/Site.php [hunk 1]
```diff
@@ -178,8 +178,6 @@ protected function getApp()
             $config->set('mail.username', $settings->get('mail_username'));
             $config->set('mail.password', $settings->get('mail_password'));
 
-            $app->register(\Flarum\BusServiceProvider::class);
-
```
Comment (franzliedke) [inline src/Foundation/Site.php]: No replacement?

**Comment by franzliedke** [comment]
@luceos Can you take care of [my comment here](https://github.com/flarum/core/commit/ea4d889b76ba03ddf55a77d6333c0a8352b9b854#commitcomment-24825880)? This can only be done when using Laravel 5.5. :)

File: src/Foundation/Site.php (hunk 1 above)
Comment (luceos) [reply src/Foundation/Site.php]: This is a WIP, I was busy stripping down flarum/flarum to get it to work with flarum/core 👍

Comment (franzliedke) [reply src/Foundation/Site.php]: Not sure what you mean, sorry. Does that mean you will revert this before finishing the PR?

**Comment by luceos** [comment]
@franzliedke [this](https://github.com/flarum/core/pull/1252#issuecomment-334926692) `@json` added.

Moving forward with this work.

File: src/Foundation/Site.php (hunk 1 above)
Comment (luceos) [reply src/Foundation/Site.php]: Fixed. Re-added the ability to auto forward Commands to their respective Handler classes.

**Comment by luceos** [comment]
This deprecates `Symfony\Component\Translation\TranslatorInterface` in our app. It is no longer bound and needs to be replaced with `Illuminate\Contracts\Translation\Translator`, you can also use `translator`.

File: src/Foundation/Site.php (hunk 1 above)
Comment (luceos) [reply src/Foundation/Site.php]: @franzliedke see https://github.com/flarum/core/pull/1252/commits/72c232d5a388479eaf712939ffeeb19bec848a19

**Comment by luceos** [comment]
Used this in my flarum/flarum after cloning that repo's `next-back` branch:

```json
    "require": {
        "flarum/core": "dev-next-back-5.5 as 0.1.0-beta.99",
        "flarum/flarum-ext-english": "dev-next-back"
    },
```

**Comment by luceos** [comment]
@flarum/core This is ready for initial review. It has been tested with:

- flarum/flarum at next-back
- flarum/flarum-ext-english at next-back

Nothing else. I went to some of the steps:

- create user
- create post
- access admin
- enable extension

In general I've noticed nothing strange.

File: src/Locale/LocaleManager.php [hunk 2]
```diff
@@ -11,12 +11,12 @@
 
 namespace Flarum\Locale;
 
-use Symfony\Component\Translation\Translator as SymfonyTranslator;
+use Illuminate\Contracts\Translation\Translator;
 
 class LocaleManager
```
Comment (tobscure) [inline src/Locale/LocaleManager.php]: LocaleManager::getTranslator docblock is still returning SymfonyTranslator.

AbstractFrontend::addTranslations depends on the SymfonyTranslator::getCatalogue, which is not present in the Illuminate contract.

I think LocaleManager needs to require a Flarum\Locale\Translator instance directly, rather than the Illuminate contract - that way we know it fulfils both the Illuminate and Symfony contracts.

File: src/Locale/LocaleServiceProvider.php [hunk 3]
```diff
@@ -41,19 +40,17 @@ public function register()
...
 
             return $translator;
         });
         $this->app->alias('translator', Translator::class);
-        $this->app->alias('translator', TranslatorInterface::class);
+        $this->app->alias('translator', TranslatorContract::class);
     }
 
-    private function getDefaultLocale()
+    private function getDefaultLocale(): string
```
Comment (tobscure) [inline src/Locale/LocaleServiceProvider.php]: A return type! That tickles 😊

File: src/Locale/Translator.php [hunk 4]
```diff
@@ -28,7 +29,6 @@ public function getCatalogue($locale = null)
         } else {
             $this->assertValidLocale($locale);
         }
-
```
Comment (tobscure) [inline src/Locale/Translator.php]: ?

File: src/Post/RegisteredTypesScope.php [hunk 5]
```diff
@@ -13,9 +13,9 @@
 
 use Illuminate\Database\Eloquent\Builder;
 use Illuminate\Database\Eloquent\Model;
-use Illuminate\Database\Eloquent\ScopeInterface;
+use Illuminate\Database\Eloquent\Scope;
 
-class RegisteredTypesScope implements ScopeInterface
+class RegisteredTypesScope implements Scope
```
Comment (tobscure) [inline src/Post/RegisteredTypesScope.php]: Scopes have changed a bit in 5.2. See https://laravel.com/docs/5.3/upgrade#upgrade-5.2.0 > Eloquent > Global Scopes.

I think we can get rid of the `remove` method, and all of the weird whereIndex/bindingIndex/bindingCount stuff that makes it possible.

Please test post type functionality after doing this to make sure it works.

File: src/Locale/LocaleServiceProvider.php (hunk 3 above)
Comment (luceos) [reply src/Locale/LocaleServiceProvider.php]: Wasn't really sure, but since we're adopting php 7.0+ and this method really only returns a string it seemed appropriate. Also because it seems to become an industry standard to get of unnecessary phpdoc's completely when type hinting do the same. What's your opinion or preference?

File: src/Locale/Translator.php (hunk 4 above)
Comment (luceos) [reply src/Locale/Translator.php]: Most likely happened during debugging as the Translator instance wasn't being resolved correctly.

File: src/Locale/LocaleServiceProvider.php (hunk 3 above)
Comment (tobscure) [reply src/Locale/LocaleServiceProvider.php]: Yes let's go for it! I only meant that it felt funny, but in a good kind of way 😆

File: src/Post/RegisteredTypesScope.php (hunk 5 above)
Comment (luceos) [reply src/Post/RegisteredTypesScope.php]: Done. Tested.

File: src/Locale/LocaleServiceProvider.php (hunk 3 above)
Comment (luceos) [reply src/Locale/LocaleServiceProvider.php]: Done, patched the whole class.

File: src/Foundation/Site.php [hunk 6]
```diff
@@ -166,6 +167,8 @@ protected function getApp()
         $app->register(ViewServiceProvider::class);
         $app->register(ValidationServiceProvider::class);
 
+        $app->register(BusProvider::class);
```
Comment (tobscure) [inline src/Foundation/Site.php]: No need for the `BusProvider` alias?

File: src/Api/Controller/SendConfirmationEmailController.php [hunk 7]
```diff
@@ -43,17 +43,17 @@ class SendConfirmationEmailController implements ControllerInterface
...
 
     /**
      * @param \Flarum\Settings\SettingsRepositoryInterface $settings
      * @param Mailer $mailer
      * @param UrlGenerator $url
-     * @param TranslatorInterface $translator
+     * @param Translator $translator
      */
-    public function __construct(SettingsRepositoryInterface $settings, Mailer $mailer, UrlGenerator $url, TranslatorInterface $translator)
+    public function __construct(SettingsRepositoryInterface $settings, Mailer $mailer, UrlGenerator $url, Translator $translator)
```
Comment (tobscure) [inline src/Api/Controller/SendConfirmationEmailController.php]: Can we please undo these changes and stick with injecting the Symfony TranslatorInterface. I think we want that to be our primary, implementation of the Illuminate contract is secondary just so the validator still works.

**Comment by tobscure** [comment]
@luceos Symfony 4.0 is out!

File: src/Foundation/Site.php (hunk 6 above)
Comment (luceos) [reply src/Foundation/Site.php]: Nope, we load both Illuminate BusServiceProvider and Flarum BusServiceProvider. Or we call the Illuminate from inside our own.
